openedx/core/djangoapps/credentials_service/models.py

```diff
@@ -1,3 +1,280 @@
+    """
+    Returns path for credentials templates file assets.
+
+    Arguments:
+        instance(CertificateTemplateAsset): CertificateTemplateAsset object
+        filename(str): file to upload
+
+    Returns:
+        Path to asset.
+    """
+
+    return 'certificate_template_assets/{id}/{filename}'.format(id=instance.id, filename=filename)
+
+
+def signatory_assets_path(instance, filename):
+    """
+    Returns path for signatory assets.
+
+    Arguments:
+        instance(Signatory): Signatory object
+        filename(str): file to upload
+
+    Returns:
+        Path to asset.
+    """
+    return 'signatories/{id}/{filename}'.format(id=instance.id, filename=filename)
+
+
+def validate_image(image):
+    """
+    Validates that a particular image is small enough.
+    """
+    if image.size > (250 * 1024):
+        raise ValidationError(_('The image file size must be less than 250KB.'))
+
+
+def validate_course_key(course_key):
+    """
+    Validate the course_key is correct.
+    """
+    try:
+        CourseKey.from_string(course_key)
+    except InvalidKeyError:
+        raise ValidationError(_("Invalid course key."))
+
+
+class SiteConfiguration(models.Model):
+    """
+    Custom Site model for custom sites/microsites.
+    """
+    site = models.OneToOneField('sites.Site', null=False, blank=False)
+    lms_url_root = models.URLField(
+        verbose_name=_('LMS base url for custom site/microsite'),
+        help_text=_("Root URL of this site's LMS (e.g. https://courses.stage.edx.org)"),
+        null=False,
+        blank=False
+    )
+    theme_scss_path = models.CharField(
+        verbose_name=_('Path to custom site theme'),
+        help_text=_('Path to scss files of the custom site theme'),
+        max_length=255,
+        null=False,
+        blank=False
+    )
+
+    def __unicode__(self):
+        """Unicode representation. """
+        return 'Site Configuration {site}'.format(
+            site=self.site
+        )
+
+
+class AbstractCredential(TimeStampedModel):
+    """
+    Abstract Credentials configuration model.
+    """
+    site = models.ForeignKey(Site)
+    is_active = models.BooleanField(default=False)
+
+    class Meta(object):
+        abstract = True
+
+
+class Signatory(TimeStampedModel):
+    """
+    Signatory model to add certificate signatories.
+    """
+    name = models.CharField(max_length=255)
+    title = models.CharField(max_length=255)
+    image = models.ImageField(
+        help_text=_(
+            'Image must be square PNG files. The file size should be under 250KB.'
+        ),
+        upload_to=signatory_assets_path,
+        validators=[validate_image]
+    )
+
+    def __unicode__(self):
+        """Unicode representation. """
+        return 'Signatory {name}, {title}'.format(
+            name=self.name,
+            title=self.title,
+        )
+
+    def save(self, *args, **kwargs):
+        """
+        The database will not enforce a primary key before saving the
```

Comment:
Here is a more descriptive explanation: A primary key/ID will not be assigned until the model is written to the database. Given that our file path relies on this ID, save the model initially with no file. After the initial save, update the file and save again. All subsequent saves will write to the database only once.